R/chemFormula.R

```diff
@@ -0,0 +1,127 @@
+#' @title Parse a chemical formula into a named vector
+#'
+#' @description
+#'
+#' `formula2list` This function parses a chemical into a named vector
+#'
+#' @param x `character` Single string with chemical
+#'
+#' @return `numeric` A named list containing the parsed chemical formula
+#'
+#' @author Michael Witting
+#' 
+#' @import stringr str_extract_all str_extract
+#'
+#' @export
+#'
+#' @examples
+#'
+#' formula2list("C6H12O6")
+#' formula2list("C11H12N2O2")
+formula2list <- function(x) {
+  
+  # regex pattern to isolate all elements
+  element_pattern <- "([A][cglmrstu]|[B][aehikr]?|[C][adeflmnorsu]?|[D][bsy]|[E][rsu]|[F][elmr]?|[G][ade]|[H][efgos]?|[I][nr]?|[K][r]?|[L][airuv]|[M][cdgnot]|[N][abdehiop]?|[O][gs]?|[P][abdmortu]?|[R][abefghnu]|[S][bcegimnr]?|[T][abcehilms]|[U]|[V]|[W]|[X][e]|[Y][b]?|[Z][nr])([0-9]*)"
+  
+  # extract all matching pattern
+  regexMatch <- stringr::str_extract_all(x, element_pattern)
+  
+  # get individual elements and their count
+  elements <- stringr::str_extract(regexMatch[[1]], "[aA-zZ]+")
+  numbers <- as.numeric(stringr::str_extract(regexMatch[[1]], "[0-9]+"))
+  
+  # replace NAs with 1 for elements which have a count of one
+  numbers[is.na(numbers)] <- 1
+  
+  # create named vector for return
+  formula_list <- numbers
+  names(formula_list) <- elements
+  
+  # remove atoms that might have a count of 0
+  formula_list <- formula_list[which(formula_list >= 0)]
+  
+  formula_list
+}
+
+#' @title Create chemical formula from a named vector
+#'
+#' @description
+#'
+#' `list2formula` This function creates a chemical formula from a parsed list
+#'
+#' @param x `character` A named vector containing the number of elements
+#'
+#' @return `character` Single string with the chemical formula
+#'
+#' @author Michael Witting
+#' 
+#' @import stringr str_extract_all str_extract
+#'
+#' @export
+#'
+#' @examples
+#' 
+#' elements <- c("C" = 6, "H" = 12, "O" = 6)
+#' list2formula(elements)
+list2formula <- function(x) {
+  
```

Comment:
maybe add a check if `x` does not have names. Something like `if (is.null(names(x))) stop("'x' should be a named vector")`.